File: old_trunk/cost/iona/orbacus/java/poa/TestDSI_impl.java
// **********************************************************************
//
// Copyright (c) 2000
// Object Oriented Concepts, Inc.
// Billerica, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

package test.poa;

import org.omg.CORBA.*;
import org.omg.PortableServer.*;

class TestDSI_impl extends org.omg.PortableServer.DynamicImplementation
{
    //
    // From TestBase (no multiple inheritance)
    //
    public static void
    TEST(boolean expr)
    {
	if(!expr)
	    throw new test.common.TestException();
    }

    protected ORB orb_;
    protected POA poa_;
    protected org.omg.PortableServer.Current current_;
    protected String name_;
    protected boolean compare_;

    TestDSI_impl(ORB orb, String name, boolean compare)
    {
	org.omg.CORBA.Object currentObj = null;

	try
	{
	    currentObj = orb.resolve_initial_references("POACurrent");
	}
	catch(org.omg.CORBA.ORBPackage.InvalidName ex)
	{
	}

	TEST(currentObj != null);
	current_ = org.omg.PortableServer.CurrentHelper.narrow(currentObj);
	TEST(current_ != null);
    }

    TestDSI_impl(ORB orb, POA poa)
    {
	orb_ = orb;
	poa_ = poa;
	name_ = "";

	org.omg.CORBA.Object currentObj = null;

	try
	{
	    currentObj = orb.resolve_initial_references("POACurrent");
	}
	catch(org.omg.CORBA.ORBPackage.InvalidName ex)
	{
	}

	TEST(currentObj != null);
	current_ = org.omg.PortableServer.CurrentHelper.narrow(currentObj);
	TEST(current_ != null);
    }

    static final String[] interfaces_ =
    {
        "IDL:Test:1.0"
    };

    public String[]
    _all_interfaces(POA poa, byte[] oid)
    {
        return interfaces_;
    }

    public boolean
    _is_a(String id)
    {
	if(id.equals("IDL:Test:1:0"))
	    return true;
	return super._is_a(id);
    }

    public void
    invoke(ServerRequest request)
    {
	String name = request.operation();

	if(name.equals("aMethod"))
	{
	    NVList list = orb_.create_list(0);
	    request.arguments(list);

	    if(compare_)
	    {
		byte[] oid = null;
		try
		{
		    oid = current_.get_object_id();
		}
		catch(org.omg.PortableServer.CurrentPackage.NoContext ex)
		{
		    throw new RuntimeException();
		}
		String oidString = new String(oid);

		TEST(oidString.equals(name_));
	    }

	    return;
	}

	System.err.println("DSI implementation: unknown operation: " + name);

	NVList list = orb_.create_list(0);
	request.arguments(list);

	Any exAny = orb_.create_any();
	BAD_OPERATIONHelper.insert(exAny, new org.omg.CORBA.BAD_OPERATION());
	request.set_exception(exAny);
    }

    public POA
    _default_POA()
    {
	if(poa_ != null)
	    return poa_;
	return super._default_POA();
    }
}

File: old_trunk/cost/iona/orbacus/cpp/poa/TestDSI_impl.cpp
// **********************************************************************
//
// Copyright (c) 2001
// Object Oriented Concepts, Inc.
// Billerica, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

#include <includes.h>

#include <Test_skel.h>

#include <TestDSI_impl.h>

using namespace CORBA;
using namespace PortableServer;

TestDSI_impl::TestDSI_impl(ORB_ptr orb, const char* name, Boolean compare)
    : orb_(CORBA::ORB::_duplicate(orb)), name_(name), compare_(compare)
{
    Object_var currentObj;

    try
    {
        currentObj = orb -> resolve_initial_references("POACurrent");
    }
    catch(const ORB::InvalidName&)
    {
    }

    TEST(!is_nil(currentObj));
    current_ = PortableServer::Current::_narrow(currentObj);
    TEST(!is_nil(current_));
}

TestDSI_impl::TestDSI_impl(ORB_ptr orb, POA_ptr poa)
    : orb_(CORBA::ORB::_duplicate(orb)), poa_(POA::_duplicate(poa)),
      name_((const char*)""), compare_(false)
{
    Object_var currentObj;

    try
    {
        currentObj = orb -> resolve_initial_references("POACurrent");
    }
    catch(const ORB::InvalidName&)
    {
    }

    TEST(!is_nil(currentObj));
    current_ = PortableServer::Current::_narrow(currentObj);
    TEST(!is_nil(current_));
}

CORBA::RepositoryId
TestDSI_impl::_primary_interface(const PortableServer::ObjectId&,
				 PortableServer::POA_ptr)
{
    return string_dup("IDL:Test:1.0");
}

CORBA::Boolean
TestDSI_impl::_is_a(const char* id)
    throw(CORBA::SystemException)
{
    if(strcmp(id, "IDL:Test:1.0") == 0)
	return true;

    return PortableServer::ServantBase::_is_a(id);
}

void
TestDSI_impl::invoke(CORBA::ServerRequest_ptr request)
    throw()
{
    String_var name = request -> operation();

    if(strcmp(name, "aMethod") == 0)
    {
	NVList_ptr list;
	orb_ -> create_list(0, list);
	request -> arguments(list);

	if(compare_)
	{
	    
	    ObjectId_var oid = current_ -> get_object_id();
	    String_var oidString = ObjectId_to_string(oid);
	    
	    TEST(strcmp(oidString, name_) == 0);
	}
	
	return;
    }

    OB_ERROR("DSI implementation: unknown operation: " << name);

    NVList_ptr list;
    orb_ -> create_list(0, list);
    request -> arguments(list);
    
    Any exAny;
    exAny <<= new BAD_OPERATION();
    request -> set_exception(exAny);
}

PortableServer::POA_ptr
TestDSI_impl::_default_POA()
{
    if(!is_nil(poa_))
        return POA::_duplicate(poa_);
    return ServantBase::_default_POA();
}
